doc/specs/#476 - Pinned packages.md

```diff
@@ -0,0 +1,186 @@
+---
+author: Easton Pillay jedieaston/easton@planeteaston.com
+created on: 2022-01-25
+last updated: 2022-02-02
+issue id: 476
+---
+
+# Pinning a package
+
+For [#476](https://github.com/microsoft/winget-cli/issues/476)
+
+## Abstract
+
+This spec describes the functionality behind "pinning" a package (freezing a package at a certain version so that it will be not be automatically upgraded via `winget upgrade --all`).
+
+## Inspiration
+
+This functionality is inspired by functionality in other package managers, as well as community feedback. Many packages (software development tools and libraries commonly, but other software as well) introduce breaking changes that users may not want integrate into their workflow quite yet.
+
+## Solution Design
+
+A table of packages (by Name, Version and Product Code if available) that are currently pinned will be kept in the local tracking catalog. To be repository independent, these values should be correlated from local sources (Add and Remove Programs, Appx, etc). This also means that the user should be able to pin packages that were not installed via the Windows Package Manager, so that they won't be upgraded via `winget upgrade --all`. 
+
+
+## UI/UX Design
+
+Several commands will need to be added to support pinning packages. Where applicable, equivalent names (or new arguments) for PowerShell cmdlets will also be given.
+
+The main interface for the pinning feature will be a new command, `winget pin`, with a couple of options:
+
+
+```
+winget pin
+PowerShell: Get-WinGetPinnedPackage
+```
+
+Running `winget pin` with no arguments will show a table of currently pinned packages:
+
+```
+Name                                                 Id                                Version    
+------------------------------------------------------------------------------------------------
+Microsoft Bob                                        Microsoft.Bob                     2.35.0
+iTunes                                               Apple.iTunes                      12.0.199.4
+
+```
+
+```
+winget pin --set <package>
```

Comment:
What command would you want to use to list pinned packages? I used `winget pin` for that, since it was reminiscent of `winget upgrade`.

Reply:
That does seem consistent with upgrade.

Reply:
I don't think it would have to be a specific command; I think it can function exactly like upgrade, where `winget pin` would list the pinned packages and `winget pin <package>` would set the pin on that package

Reply:
Ah, I see. `winget pin` with no args shows the table, `winget pin <package query>` sets the pin, and `winget pin --clear <package query>` unpins. That would mean you wouldn't be able to filter the pin table, which probably doesn't happen that often but would make it inconsistent with install/list/upgrade.

Reply:
> Ah, I see. `winget pin` with no args shows the table, `winget pin <package query>` sets the pin, and `winget pin --clear <package query>` unpins. That would mean you wouldn't be able to filter the pin table, which probably doesn't happen that often but would make it inconsistent with install/list/upgrade.

That would be consistent with `upgrade`, but not `list`.  An option to allow filtering would be `winget pin <query> --list`.  Considering it is a modifying command, I think it makes sense to remain consistent with `upgrade`.  `--list` could even be added to `upgrade`, although that behavior is already equivalently exposed via `list` itself.

Reply:
Updated in latest commit. `winget pin` with no args shows the list, `winget pin <package>` pins, and `winget pin --clear <package>` removes the pin.